ops/framework.py

```diff
@@ -852,33 +852,34 @@ def __get__(self, parent, parent_type=None):
             # accessing via the class directly (e.g. MyClass.stored)
             return self
 
-        if self.attr_name is not None and self.attr_name in parent.__dict__:
-            # we probably already have the thing
-            bound = parent.__dict__[self.attr_name]
-
-            if bound is not None:
-                return bound
+        bound = parent.__dict__.get(self.attr_name)
+        if bound is not None:
+            # we already have the thing from a previous pass, huzzah
+            return bound
 
         # need to find ourselves amongst the parent's bases
         for cls in parent_type.mro():
             for attr_name, attr_value in cls.__dict__.items():
                 if attr_value is not self:
                     continue
-                if self.attr_name is not None and attr_name != self.attr_name:
+                # we've found ourselves! is it the first time?
+                if bound is not None:
                     # the StoredState instance is being stored in two different
                     # attributes -> unclear what is expected of us -> bail out
                     raise RuntimeError("StoredState shared by {0}.{1} and {0}.{2}".format(
                         cls.__name__, self.attr_name, attr_name))
-                # we've found ourselves; save where, and bind the object (and cache it)
+                # we've found ourselves for the first time; save where, and bind & cache the object
                 self.attr_name = attr_name
                 self.parent_type = cls
                 bound = BoundStoredState(parent, attr_name)
                 parent.__dict__[attr_name] = bound
-                return bound
+
+        if bound is not None:
+            return bound
```

Comment:
The `parent.__dict__[attr_name] = bound` assignment above needs to be moved inside this branch for correctness. Otherwise, this logic is raising an error, but doing exactly the same thing it would do if the logic was correct by mutating the parent. If the error is caught for whatever reason, and this is called again, the error will not be raised and the bug is now a mysterious side effect.